org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ContextCalculator.java

```diff
@@ -0,0 +1,224 @@
+	 * from which the TargetFields will be calculated.
+	 *
+	 * @param method the IMethod to be processed.
+	 */
+	public ContextCalculator(IMethod method) {
+		this.fSelectionInputType= SelectionInputType.IMETHOD;
+		this.fTargetIMethod= method;
+	}
+
+	public SelectionInputType getSelectionInputType() {
+		return fSelectionInputType;
+	}
+
+	public Selection getSelectionEditorText() {
+		return fSelectionEditorText;
+	}
+
+	public ICompilationUnit getSelectionICompilationUnit() {
+		return fSelectionICompilationUnit;
+	}
+
+	public ASTNode getOrComputeSelectionASTNode() {
+		if (fSelectionASTNode==null) {
+			calculateSelectionASTNode();
+		}
+		return fSelectionASTNode;
+	}
+
+	public IMethodBinding getOrComputeTargetIMethodBinding() {
+		if(fTargetIMethodBinding==null) {
+			calculateTargetIMethodBinding();
+		}
+		return fTargetIMethodBinding;
+	}
+
+	public IMethod getOrComputeTargetIMethod() {
+		if(fTargetIMethod==null) {
+			calculateTargetIMethod();
+		}
+		return fTargetIMethod;
+	}
+
+	public ICompilationUnit getOrComputeTargetICompilationUnit() {
+		if(fTargetICompilationUnit==null) {
+			calculateTargetICompilationUnit();
+		}
+		return fTargetICompilationUnit;
+	}
+
+	public MethodDeclaration getOrComputeTargetMethodDeclaration() throws JavaModelException {
+		if (fTargetMethodDeclaration==null) {
+			calculateMethodDeclaration();
+		}
+		return fTargetMethodDeclaration;
+	}
+
+	/**
+	 * This method calculates the selected ASTNode {@link #fSelectionASTNode}. It finds the Node
+	 * that is inside the {@link #fSelectionICompilationUnit} CompilationUnit at the
+	 * {@link #fSelectionEditorText} Selection.
+	 */
+	private void calculateSelectionASTNode() {
+		fSelectionCompilationUnit= convertICompilationUnitToCompilationUnit(fSelectionICompilationUnit);
+		fSelectionASTNode= NodeFinder.perform(fSelectionCompilationUnit, fSelectionEditorText.getOffset(), fSelectionEditorText.getLength());
+
+		if (fSelectionASTNode == null) {
+			return;
+		}
+
+		if (fSelectionASTNode.getNodeType() == ASTNode.SIMPLE_NAME) {
+			fSelectionASTNode= fSelectionASTNode.getParent();
+		} else if (fSelectionASTNode.getNodeType() == ASTNode.EXPRESSION_STATEMENT) {
+			fSelectionASTNode= ((ExpressionStatement) fSelectionASTNode).getExpression();
+		}
+	}
+
+	/**
+	 * This method calculates the {@link #fTargetIMethodBinding}. If the {@link #fSelectionASTNode}
+	 * is an instance of {@link MethodInvocation} or {@link MethodDeclaration}, it resolves the
+	 * corresponding {@link IMethodBinding} from that instance.
+	 */
+	private void calculateTargetIMethodBinding() {
+		if (getOrComputeSelectionASTNode() instanceof MethodInvocation selectionMethodInvocation) {
+			fTargetIMethodBinding= selectionMethodInvocation.resolveMethodBinding();
+		} else if (fSelectionASTNode instanceof MethodDeclaration selectionMethodDeclaration) {
+			fTargetIMethodBinding= selectionMethodDeclaration.resolveBinding();
+		}
```

Comment:
As far as I understand the code, the `calculate...` methods are always expected to return something that is not `null` by their callers (i.e., `getOrComputed...`. Otherwise anyone using the result of `getOrCompute` may retrieve a `null` value that then run into an NPE. For example, calling `calculateTargetIMethod()` will throw an NPE in case `the selection ast node is neither a `MethodInvocation` nor a `MethodDeclaration`. So there should always be a fallback logic in these cases.

Reply:
Done.
d0c8e158cf6c9df594271bea787c28e260880317